src/classes/RD2_RecurringDonation.cls

```diff
@@ -66,7 +71,28 @@ public inherited sharing class RD2_RecurringDonation {
     * @param rd Recurring Donation record
     */
     public RD2_RecurringDonation(npe03__Recurring_Donation__c rd) {
+        this(rd, rd.npe03__Donations__r);
+    }
+
+    /**
+    * @description Constructor
+    * @param rd Recurring Donation record
+    * @param opportunities Recurring Donation current and future Opportunities
+    */
+    public RD2_RecurringDonation(npe03__Recurring_Donation__c rd, List<Opportunity> opportunities) {
         this.rd = rd;
+
+        this.opportunities = opportunities == null 
```

Comment:
When the Next Donation Date is determined in the RD2 trigger on update, then `npe03__Donations__r` list is not set. Thus, it is better to store Opportunities in a variable without requiring `npe03__Recurring_Donation__c`  to be rebuilt with `npe03__Donations__r`  populated.

Reply:
I like this change a lot.  Makes the implementation much more general I think.